src/mkdocstrings/_cache.py

```diff
@@ -26,6 +32,46 @@ def download_url_with_gz(url: str) -> bytes:
         return content.read()
 
 
+def _extract_auth_from_url(url: str) -> tuple[str, dict[str, str]]:
+    """Extracts the username and password from the URL, if present, and returns the URL and the auth header."""
+    parsed_url = urllib.parse.urlparse(url)
+
+    auth_header: dict[str, str] = {}
+    if parsed_url.username:
+        # If the URL contains a username, we assume the user is trying to authenticate
+        auth_header = _create_auth_header(user_env_var=parsed_url.username, pwd_env_var=parsed_url.password)
+        # Remove the auth part from the URL
+        url = parsed_url._replace(netloc=parsed_url.hostname or "").geturl()
+    return url, auth_header
+
+
+def _create_auth_header(user_env_var: str, pwd_env_var: Union[str, None] = None) -> dict[str, str]:
+    """Creates the Authorization header for basic authentication."""
+    user = _get_environment_variable(user_env_var)
+
+    if pwd_env_var is None:
+        # If password is not provided, we assume that the user is using a token
+        log.debug("Using bearer token for authentication")
+        return {"Authorization": f"Bearer {user}"}
+
+    # Else, we assume that the user is using user:password
+    pwd = _get_environment_variable(pwd_env_var)
+    log.debug("Using basic authentication")
+    return {"Authorization": f"Basic {user}:{pwd}"}
+
+
+def _get_environment_variable(env_var: str) -> str:
+    """Extracts the environment variable name from env_var and returns its value."""
+    match = ENV_VAR_PATTERN.match(env_var)
+    if not match:
+        raise ValueError("URL authentication must be specified as environment variables")
```

Comment:
Do we actually want to prevent users from hardcoding credentials? They're responsible adults.

In other words, couldn't we refactor this to just regex-replace environment variables found in the username and password parts? This way we would directly support `https://prefix${MY_USERNAME}suffix:prefix${MY_PASSWORD}suffix@example.com/path/objects.inv`, which I think could be asked for later anyway.

There's the case where `${VALID_NAME}` appears in a hardcoded username/password though. We probably don't want to output false-positive warnings, but we also want to warn users when they do want to use an env var, but it's unset. Maybe we can just log such warnings/hints when the URL fails to download, a bit later in the process. WDYT? Maybe I'm just over-complicating things too early.

Reply:
I can take it from there if you want.

Reply:
I'm working on the first part of your suggestion (regex replacing env vars). I found the implementation that [PDM uses](https://github.com/pdm-project/pdm/blob/ea39d366a3d57b46517a1617a96c96a3fccb7d12/src/pdm/utils.py#L252) (thanks to your comment), which seems rather neat.

In regards to your second comment on `${VALID_BUT_UNSET}`, I think it makes sense to log it as a warning. It is probably quite unlikely that e.g. the password happens to match a pattern of `${...}`, but I see no reason to fail hard in that case.

Reply:
OK, lets go with an early warning :slightly_smiling_face: If people complain we can always move the warning to a later time in the process, as I suggested.

Reply:
Sure, feel free to add a comment above the relevant functions :slightly_smiling_face: